File: src/python/finished/maximum_subarray_sum_with_length_divisible_by_k.py
from typing import List

# https://leetcode.com/problems/maximum-subarray-sum-with-length-divisible-by-k/

class Solution:
    def maxSubarraySum(self, nums: List[int], k: int) -> int:
        bests = [-float('inf')] * k
        best = -float('inf')
        acc = 0
        for i, n in enumerate(nums):
            acc += n
            if i >= k:
                acc -= nums[i - k]
            if i >= k - 1 :
                j = i % k
                bests[j] = max(bests[j] + acc, acc)
                best = max(best, bests[j])
        return best

File: src/rust/finished/maximum_subarray_sum_with_length_divisible_by_k.rs
impl Solution {
    pub fn max_subarray_sum(nums: Vec<i32>, k: i32) -> i64 {
        let k = k as usize;
        let mut bests = vec![i32::MIN as i64; k];
        let mut best = i64::MIN;
        let mut acc = 0i64;
        for (i, &n) in nums.iter().enumerate() {
            acc += n as i64;
            if i >= k {
                acc -= nums[i - k] as i64;
            }
            if i >= k - 1 {
                let j = i.rem_euclid(k);
                bests[j] = acc.max(bests[j] + acc);
                best = best.max(bests[j])
            }
        }
        best
    }
}
